source/material_model/morency_doin.cc

```diff
@@ -73,42 +73,65 @@ namespace aspect
-          const double e2inv = std::sqrt(std::pow(second_invariant(strain_rate),2) + std::pow(min_strain_rate,2));
-          double one_over_veffv = 0.0;
-          if (temperature != 0.0)
-            one_over_veffv = 1.0 /
-                             (B * std::pow(e2inv/ref_strain_rate, -1.0+1.0/nv) * std::exp((activation_energy+activation_volume*reference_density()*g*z)/(nv*R*temperature)));
-          const double veffp = tauy * (std::pow(e2inv, -1.0+1.0/np) / std::pow(ref_strain_rate, 1.0/np));
-          // Effective viscosity = harmonic mean of diffusion and dislocation creep. Range is limited to 1e17-1e28 for stability.
-          const double veff = std::min(std::max(veff_coefficient * std::pow((1.0/veffp + one_over_veffv), -1.0), min_visc), max_visc);
-          out.viscosities[i] = veff;
-
-          double density = 0.0;
-          for (unsigned int j=0; j < volume_fractions.size(); ++j)
-            {
-              //not strictly correct if thermal expansivities are different, since we are interpreting
-              //these compositions as volume fractions, but the error introduced should not be too bad.
-              const double temperature_factor= (1.0 - thermal_expansivities[j] * (temperature - reference_T));
-              density += volume_fractions[j] * densities[j] * temperature_factor;
-            }
-          out.densities[i] = density;
+          const SymmetricTensor<2,dim> e = in.strain_rate[i];
+          // Check if strain rate is undefined
+          if (std::isnan(e[0][0]))
+            for (unsigned int x=0; x<dim; ++x)
+              for (unsigned int y=0; y<dim; ++y)
+                e[x][y] = 0.0;
+
+          const double depth = this->get_geometry_model().depth(position); // units: m
+
+          // Calculate viscosity
+          {
+            double activation_energy = 0.0;
+            for (unsigned int j=0; j < volume_fractions.size(); ++j)
+              activation_energy += volume_fractions[j] * activation_energies[j] * 1e3; // Converted to J/mol to make units work
+            double nv = 0.0;
+            for (unsigned int j=0; j < volume_fractions.size(); ++j)
+              nv += volume_fractions[j] * nvs[j];
+            double np = 0.0;
+            for (unsigned int j=0; j < volume_fractions.size(); ++j)
+              np += volume_fractions[j] * nps[j];
+
+            const double tauy = tau_0 + gamma*reference_density()*g*depth;
+
+            /**
+             * Note: The actual second invariant of strain rate is defined as $e_{00}e_{11} - e_{01}^2$,
+             * (maybe times 1/2 depending on the convention) where e is the second order strain rate
+             * tensor. The function used here reflects the equations defined in (Morency and Doin, 2004)
+             * but is otherwise unjustified. Use at your own risk.
+             */
+            const double e2inv = std::sqrt((strain_rate[0][0]*strain_rate[0][0] + strain_rate[0][1]*strain_rate[0][1])/2.0);
```

Comment:
Oops. Must have been an artifact of one of the rebases. I switched back and forth and decided on going with the strain_rate variable name because it's more descriptive, albeit verbose.